plugins/inputs/vsphere/vsan.go

```diff
@@ -0,0 +1,539 @@
+func getCmmdsMap(ctx context.Context, client *vim25.Client, clusterObj *object.ClusterComputeResource) (map[string]CmmdsEntity, error) {
+	hosts, err := clusterObj.Hosts(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("fail to get host: %v", err)
+	}
+
+	if len(hosts) == 0 {
+		return make(map[string]CmmdsEntity), nil
+	}
+
+	queries := make([]types.HostVsanInternalSystemCmmdsQuery, 2)
+	hostnameCmmdsQuery := types.HostVsanInternalSystemCmmdsQuery{
+		Type: "HOSTNAME",
+	}
+	diskCmmdsQuery := types.HostVsanInternalSystemCmmdsQuery{
+		Type: "DISK",
+	}
+	queries = append(queries, hostnameCmmdsQuery)
+	queries = append(queries, diskCmmdsQuery)
+
+	//Some esx host can be down or in maintenance mode. Hence cmmds query might fail on such hosts.
+	// We iterate until be get proper api response
+	var resp *types.QueryCmmdsResponse
+	for _, host := range hosts {
+		vis, err := host.ConfigManager().VsanInternalSystem(ctx)
+		if err != nil {
+			continue
+		}
+		request := types.QueryCmmds{
+			This:    vis.Reference(),
+			Queries: queries,
+		}
+		resp, err = methods.QueryCmmds(ctx, client.RoundTripper, &request)
+		if err == nil {
+			break
+		}
+	}
+	if resp == nil {
+		return nil, fmt.Errorf("all hosts fail to query cmmds")
+	}
+	var clusterCmmds Cmmds
+
+	err = json.Unmarshal([]byte(resp.Returnval), &clusterCmmds)
+	if err != nil {
+		return nil, fmt.Errorf("fail to convert cmmds to json: %v", err)
+	}
+
+	cmmdsMap := make(map[string]CmmdsEntity)
+	for _, entity := range clusterCmmds.Res {
+		uuid := entity.UUID
+		cmmdsMap[uuid] = entity
+	}
+	return cmmdsMap, nil
+}
+
+// queryPerformance adds performance metrics to telegraf accumulator
+func (e *Endpoint) queryPerformance(ctx context.Context, vsanClient *soap.Client, clusterRef *objectRef, metrics map[string]string, cmmds map[string]CmmdsEntity, acc telegraf.Accumulator) error {
+	end := time.Now().UTC()
+
+	// We're using a fake metric key, since we only store one highwater mark per resource
+	start, ok := e.hwMarks.Get(hwMarksKeyPrefix+clusterRef.ref.Value, "generic")
+	if !ok {
+		// Look back 3 sampling periods by default
+		start = end.Add(time.Duration(e.Parent.MetricLookback) * time.Duration(-e.resourceKinds["vsan"].sampling) * time.Second)
+	}
+	e.Parent.Log.Debugf("[vSAN] Query vsan performance for time interval: %s ~ %s", start, end)
+	latest := start
+
+	for entityRefID := range metrics {
+		if !strings.HasPrefix(entityRefID, perfPrefix) {
+			continue
+		}
+		entityRefID = strings.TrimPrefix(entityRefID, perfPrefix)
+		var perfSpecs []vsantypes.VsanPerfQuerySpec
+
+		perfSpec := vsantypes.VsanPerfQuerySpec{
+			EntityRefId: fmt.Sprintf("%s:*", entityRefID),
+			StartTime:   &start,
+			EndTime:     &end,
+		}
+		perfSpecs = append(perfSpecs, perfSpec)
+
+		perfRequest := vsantypes.VsanPerfQueryPerf{
+			This:       perfManagerRef,
+			QuerySpecs: perfSpecs,
+			Cluster:    &clusterRef.ref,
+		}
+		resp, err := vsanmethods.VsanPerfQueryPerf(ctx, vsanClient, &perfRequest)
+
```

Comment:
Please group code that logically belongs together and avoid empty lines between those elements. Like in this example, do not put a newline between assigning `err` and checking it.